packages/flutter/lib/src/painting/text_linker.dart

```diff
@@ -0,0 +1,424 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'inline_span.dart';
+import 'text_span.dart';
+
+/// Signature for a function that builds an [InlineSpan] link.
+///
+/// The link displays [displayString] and links to [linkString] when tapped.
+/// These are distinct because sometimes a link may be split across multiple
+/// [TextSpan]s.
+///
+/// {@template flutter.painting.LinkBuilder.recognizer}
+/// It's necessary for the owning widget to manage the lifecycle of any
+/// [GestureRecognizer]s created in this function, such as for handling a tap on
+/// the link. See [TextSpan.recognizer] for more.
+/// {@endtemplate}
+typedef InlineLinkBuilder = InlineSpan Function(
+  String displayString,
+  String linkString,
+);
+
+/// Singature for a function that finds [TextRange]s in the given [String].
+typedef TextRangesFinder = Iterable<TextRange> Function(String text);
+
+/// Specifies a way to find and style parts of some text.
+///
+/// [TextLinker]s can be applied to some text using the [linkSpans] method.
+///
+/// [textRangesFinder] must not produce overlapping [TextRange]s.
+///
+/// {@tool dartpad}
+/// This example shows how to use [TextLinker] to link both URLs and Twitter
+/// handles in the same text.
+///
+/// ** See code in examples/api/lib/painting/text_linker/text_linker.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [LinkedText.textLinkers], which uses [TextLinker]s to allow full control
+///    over matching and building different types of links.
+///  * [LinkedText.new], which is simpler than using [TextLinker] and
+///    automatically manages the lifecycle of any [GestureRecognizer]s.
+class TextLinker {
+  /// Creates an instance of [TextLinker] with a [TextRangesFinder] and an
+  /// [InlineLinkBuilder].
+  ///
+  /// The [TextRangesFinder] must not produce overlapping matches.
+  ///
+  /// Does not manage the lifecycle of any [GestureRecognizer]s created in the
+  /// [InlineLinkBuilder], so it's the responsibility of the caller to do so.
+  /// See [TextSpan.recognizer] for more.
+  TextLinker({
+    required this.textRangesFinder,
```

Comment:
whats the use case for `textRangesFinder` vs  just a regex?

Reply:
TextRangesFinder uses a function to find matches instead of just a RegExp.  I'm imagining looking up matches in a large Map or some data structure...

Reply:
> I'm imagining looking up matches in a large Map or some data structure

Even so the regex should be enough?

Is supporting linking without a regex part of the use cases we want to support, it seems to me at that point what we are doing is helping people parse textspan into a smaller pieces given a text range. but the return data (displayString and linkString) does not seem useful....

Reply:
See [this comment](https://github.com/flutter/flutter/pull/125927#discussion_r1308039346) for an explanation of the displayString/linkString thing.  I agree users probably don't care, but I can't think of a better way to do it right now.  In the long term, @LongCatIsLooong was talking about introducing a flat string annotation data structure that would make this cleaner.

About textRangesFinder, I think I'm up for making it private for now and just exposing RegExp.  I can consider opening it up if anyone asks for it.